### ISSUE TITLE
This software can't keep credentials stored

### ISSUE THREAD

**Issue by Elvino00** [kylon/Sharedown#51]
Basically, even if I select "manual login" as login module and then enable the Chrome user data folder (as you taught me from some previous issue) the software doesn't keep credentials memorized (My university forces 2FA now). Also, after putting credentials, when it's time to play the video, there is always "playback issue" on the page and I have to re-execute at least 3 times (putting in every time the credentials) the software in order to obtain a successful download. Here are the logs. By the way, I am on Linux x64 bit (Fedora 36) at the moment.

[sharedownLog.log](https://github.com/kylon/Sharedown/files/9821559/sharedownLog.log)
[ytdlp.log](https://github.com/kylon/Sharedown/files/9821561/ytdlp.log)

EDIT: I Started another download and these are the errors I always get:

![Screenshot from 2022-10-19 16-57-53](https://user-images.githubusercontent.com/74961430/196727950-f5d5e8ab-7b29-4436-bd3c-db42a5eb7f52.png)
![Screenshot from 2022-10-19 16-58-00](https://user-images.githubusercontent.com/74961430/196727958-80b27bfe-42a7-48e7-8e82-7ba95221038f.png)

**Comment by kylon** [comment]
Logs are clean, something else happening way before due to your uni login changes

You must select yes on first login when it asks to keep credentials or to remember the session otherwise it will not survive the next download

About the error, should be just a missing null check but again i cant reproduce with my uni, just finished 40 tests and no error so i expect to be something specific

I ll post a test version

**Comment by Elvino00** [comment]
> Logs are clean, something else happening way before due to your uni login changes
> 
> You must select yes on first login when it asks to keep credentials or to remember the session otherwise it will not survive the next download
> 
> About the error, should be just a missing null check but again i cant reproduce with my uni, just finished 40 tests and no error so i expect to be something specific
> 
> I ll post a test version

I always select yes when it asks to keep credentials and tick the box "don't show this message again"

**Comment by kylon** [comment]
It is possible it is now clearing data when you close the browser, in this case you need to use the login module + remember login credentials

It uses the system keyring, tested with gnome keyring but it should support other most common ones

**Comment by kylon** [comment]
Ah 2fa, prevous reply is wrong

Mh if it is clearing data on close i think a way would be to keep the browser open unless there is a puppeteer option for this
I dont see any other way atm for cases like this

**Comment by Elvino00** [comment]
> Ah 2fa, prevous reply is wrong
> 
> Mh if it is clearing data on close i think a way would be to keep the browser open unless there is a puppeteer option for this I dont see any other way atm for cases like this

How do i keep the test browser open if it automatically closes when download starts?

**Comment by kylon** [comment]
you can't, i need to add a setting for this, i ll add it to the test version

**Comment by Elvino00** [comment]
> you can't, i need to add a setting for this, i ll add it to the test version

Oh, ok then. I'm ready to test when you release it.

**Comment by kylon** [comment]
[Sharedownt1.zip](https://github.com/kylon/Sharedown/files/9847108/Sharedownt1.zip)

**Comment by kylon** [comment]
you say `reopened the browser`, does it mean you close its window while downloading?
the option is meant to keep the window open so session is not cleared

**Comment by Elvino00** [comment]
> you say `reopened the browser`, does it mean you close its window while downloading? the option is meant to keep the window open so session is not cleared

No, not while downloading. I closed it when the downloads where completed.

**Comment by kylon** [comment]
ok, if you close the browser window, since your login session does not survive, when a new browser is launched, it will ask for credentials again
so the workaround is to keep the browser open while you are using sharedown, don't close it on first batch completed, close it when you exit sharedown (before or after)

data folder is probably not needed in this case, you can just use login module, since session is lost on close

this is not an ideal solution but i have no way to test and no 2fa settings
if the same happens with a standard browser, not launched by sharedown, ie you login, 2fa etc.. and then close the browser and once reopened you must redo the login process, it is the way set by your uni, so probably this is the only? workaround, otherwise i would need help from someone that is willing to work on this and send pr

**Comment by Elvino00** [comment]
> ok, if you close the browser window, since your login session does not survive, when a new browser is launched, it will ask for credentials again so the workaround is to keep the browser open while you are using sharedown, don't close it on first batch completed, close it when you exit sharedown (before or after)
> 
> data folder is probably not needed in this case, you can just use login module, since session is lost on close
> 
> this is not an ideal solution but i have no way to test and no 2fa settings if the same happens with a standard browser, not launched by sharedown, ie you login, 2fa etc.. and then close the browser and once reopened you must redo the login process, it is the way set by your uni, so probably this is the only? workaround, otherwise i would need help from someone that is willing to work on this and send pr

First, I have changed the executable path because the chromium based internal browser gives me an error about "send input" or something like that (I'll send the logs). I am using google-chrome-stable now, using Unina login module and I've disabled the setting about chrome data folder and selected the "remeber login credentials option". I'll keep you posted on what happens.

**Comment by kylon** [comment]
it gets stuck after one download for every download? and it still asks for 2fa even if the browser is open?

i ll add lot more log prints and see if it can catch something useful

**Comment by Elvino00** [comment]
> it gets stuck after one download for every download? and it still asks for 2fa even if the browser is open?
> 
> i ll add lot more log prints and see if it can catch something useful

I start one download, the browser opens and asks for 2fa, the video is played and the download starts. Then, when this download is completed and it's time for the second one, I get an error I can't remember but all downloads stop and I have to close the browser, and close sharedown too cause if I try to start manually the second download I get an error that it can't find a valid link.

**Comment by Elvino00** [comment]
> it gets stuck after one download for every download? and it still asks for 2fa even if the browser is open?
> 
> i ll add lot more log prints and see if it can catch something useful

i guess you didn't catch anything, right?

**Comment by Elvino00** [comment]
> my uni setup is different, i ll post a version with more logs but had no time, probably this week

oh sorry, I'll wait then.

**Comment by kylon** [comment]
[Sharedownt2.zip](https://github.com/kylon/Sharedown/files/9954005/Sharedownt2.zip)

**Comment by Elvino00** [comment]
> [Sharedownt2.zip](https://github.com/kylon/Sharedown/files/9954005/Sharedownt2.zip)

I've put three items in download, so I put my credentials and sms (click "yes" to the pop up asking if i wanna stay signed in) and the first video is played. The download starts smoothly. When the first one is done, another tab is opened where the second video is opened but suddendly i get these two errors:

first: 
https://imgur.com/WHKo0Ly

second (after clicking on "ok" ) :
https://imgur.com/zRT0ssZ



These are the logs:
[sharedownLog.log](https://github.com/kylon/Sharedown/files/9954463/sharedownLog.log)
[ytdlp.log](https://github.com/kylon/Sharedown/files/9954464/ytdlp.log)


Basically I'm unable to download automatically more videos (and I keep the browser opened because of 2FA. Obviously I ticked the option in the sharedown)

**Comment by kylon** [comment]
use manual login

**Comment by Elvino00** [comment]
> use manual login

Just tried using the manual login. It still gives me the same errors in the pictures.  I can't do more than one download at the same time

[sharedownLog.log](https://github.com/kylon/Sharedown/files/9982698/sharedownLog.log)
[ytdlp.log](https://github.com/kylon/Sharedown/files/9982700/ytdlp.log)

**Comment by kylon** [comment]
[Sharedownt4.zip](https://github.com/kylon/Sharedown/files/9995017/Sharedownt4.zip)

**Comment by kylon** [comment]
[Sharedownt6.zip](https://github.com/kylon/Sharedown/files/9997784/Sharedownt6.zip)

try this with bundled chromium, not custom, it should now work and hopefully solves this issue

**Comment by Elvino00** [comment]
> [Sharedownt4.zip](https://github.com/kylon/Sharedown/files/9995017/Sharedownt4.zip)

I tried this and it doesn't work but I've just seen this "Sharedown6.zip" file so I'll check that instead

**Comment by Elvino00** [comment]
> [Sharedownt6.zip](https://github.com/kylon/Sharedown/files/9997784/Sharedownt6.zip)
> 
> try this with bundled chromium, not custom, it should now work and hopefully solves this issue

How do i reset the program to default settings?

**Comment by kylon** [comment]
you can delete sett file, in File -> open logs folder, go to up folder and delete the file

**Comment by Elvino00** [comment]
> you can delete sett file, in File -> open logs folder, go to up folder and delete the file

The software says it can't find the chromium executable and I0ve set the program to use the built-in one

[sharedownLog.log](https://github.com/kylon/Sharedown/files/9997863/sharedownLog.log)
[ytdlp.log](https://github.com/kylon/Sharedown/files/9997864/ytdlp.log)


![image](https://user-images.githubusercontent.com/74961430/201535602-8a644b0d-41d4-447c-bbf3-9006eb6bd592.png)

**Comment by kylon** [comment]
is the folder node_modules/puppeteer/chrome present? if not, the modules install failed or you are using an old node_modules folder
please delete your node_modules folder and .lock file and run `yarn install` or `npm i` to install modules

**Comment by kylon** [comment]
checked the zip, a file was missing
[Sharedownt61.zip](https://github.com/kylon/Sharedown/files/9997880/Sharedownt61.zip)

**Comment by Elvino00** [comment]
> checked the zip, a file was missing [Sharedownt61.zip](https://github.com/kylon/Sharedown/files/9997880/Sharedownt61.zip)

ok, I'm retrying with this. Also, just to say, i compile with npm

**Comment by kylon** [comment]
are you setting a ms id in sharedown settings?
is your login page changed recently?

try with no auto login, leave ms id empty in settings

**Comment by kylon** [comment]
oh wait, that is the 2nd download?

**Comment by Elvino00** [comment]
> oh wait, that is the 2nd download?
> 
> EDIT:
> 
> if `keep browser open` is enabled, please disable it and retry

yes, it was the second

**Comment by kylon** [comment]
try without the new experimental setting and enable auto login

**Comment by Elvino00** [comment]
> try without the new experimental setting and enable auto login

What happens now is that when the second download starts I get these two errors:
[first](https://user-images.githubusercontent.com/74961430/201539281-41340ef1-ebce-4971-88bd-d7063e8d07aa.png)
[second](https://imgur.com/zRT0ssZ)

[sharedownLog.log](https://github.com/kylon/Sharedown/files/9997996/sharedownLog.log)
[ytdlp.log](https://github.com/kylon/Sharedown/files/9997997/ytdlp.log)

**Comment by kylon** [comment]
[Sharedownt7.zip](https://github.com/kylon/Sharedown/files/9998012/Sharedownt7.zip)

**Comment by Elvino00** [comment]
> [Sharedownt7.zip](https://github.com/kylon/Sharedown/files/9998012/Sharedownt7.zip)

I tried to start the first download and i get this [error](https://user-images.githubusercontent.com/74961430/201540262-b76c059b-176c-41b4-b15b-ae6ddea8f73d.png).


[sharedownLog.log](https://github.com/kylon/Sharedown/files/9998030/sharedownLog.log)
[ytdlp.log](https://github.com/kylon/Sharedown/files/9998031/ytdlp.log)

**Comment by kylon** [comment]
[Sharedownt8.zip](https://github.com/kylon/Sharedown/files/9998461/Sharedownt8.zip)

**Comment by Elvino00** [comment]
> [Sharedownt8.zip](https://github.com/kylon/Sharedown/files/9998461/Sharedownt8.zip)

I tried this build and i was able to do three consecutive downloads. Here's what I've done: I used UNINA authentication module so the only thing to input was the 2FA code. I've left the option to keep the browser window opened (the experimental setting you were referring to ?) off, as you suggested.

**Comment by kylon** [comment]
ok, so it should be fixed now

about the original issue, credentials, can you try again to enable chrome data folder and disable auto login to keep credentials?

if again it forgets them, i ll post a new test with fixed keep browser open option

**Comment by Elvino00** [comment]
> ok, so it should be fixed now
> 
> about the original issue, credentials, can you try again to enable chrome data folder and disable auto login to keep credentials?
> 
> if again it forgets them, i ll post a new test with fixed keep browser open option

I didn't understand if I have to keep enabled [this option](https://user-images.githubusercontent.com/74961430/201697820-8b74b681-6b54-48f0-b80a-39aeec4b94e7.png)

**Comment by kylon** [comment]
no, that one is disabled

**Comment by Elvino00** [comment]
> ok, so it should be fixed now
> 
> about the original issue, credentials, can you try again to enable chrome data folder and disable auto login to keep credentials?
> 
> if again it forgets them, i ll post a new test with fixed keep browser open option

I tried to do as you told but I got [this error](https://user-images.githubusercontent.com/74961430/201700858-3ff67d3c-cd20-4871-96b9-7946f4b68fe7.png)

The following settings are enabled:

[Image 1](https://user-images.githubusercontent.com/74961430/201701364-8e3e8d7f-0823-4bb3-bb38-3b3acb56b56d.png)
[Image 2](https://user-images.githubusercontent.com/74961430/201701368-12c82c1a-a164-4d13-acb2-26bf192336f4.png)
[Image 3](https://user-images.githubusercontent.com/74961430/201701372-b9e8a029-da40-4d8b-bc16-d2f50d0faf45.png)

These are the logs:


[sharedownLog.log](https://github.com/kylon/Sharedown/files/10004542/sharedownLog.log)
[ytdlp.log](https://github.com/kylon/Sharedown/files/10004543/ytdlp.log)

**Comment by kylon** [comment]
uh, though that was fixed, so it only works if you enable auto login and no chrome folder?

**Comment by kylon** [comment]
[Sharedownt9.zip](https://github.com/kylon/Sharedown/files/10004708/Sharedownt9.zip)

keep browser open should be working here

disable auto login
disable chrome data folder
enable keep browser open

on first download it will ask for credentials, and the browser window will never close

**Comment by Elvino00** [comment]
after downloading the first one I got [this error](https://user-images.githubusercontent.com/74961430/201715890-acae0e10-ca09-4c07-bf65-0e0c8ad2e18b.png) but I am connnected to the network. I refreshed the page and then I got puppeteer error. I only have enabled "keep the browser windows opened". No chrome folder or remember login credential and login module is manual




[sharedownLog.log](https://github.com/kylon/Sharedown/files/10005096/sharedownLog.log)
[ytdlp.log](https://github.com/kylon/Sharedown/files/10005097/ytdlp.log)

**Comment by kylon** [comment]
[Sharedownt10.zip](https://github.com/kylon/Sharedown/files/10005455/Sharedownt10.zip)

**Comment by Elvino00** [comment]
> 

I've authenticated and I've been able to do 3 consecutive downloads. Basic authentication, haven't tested Unina yet. All of this with the option to keep the browser window opened and i had to authenticate only once, for the rest, seems that the downloads went all smoothly without the need to intervene. 

If you want to check the logs, here they are:

[sharedownLog.log](https://github.com/kylon/Sharedown/files/10006012/sharedownLog.log)
[ytdlp.log](https://github.com/kylon/Sharedown/files/10006013/ytdlp.log)

**Comment by kylon** [comment]
nothing really changed, just added logs
previous log was fine too, it got the data but for unknown reasons it failed to grab the video title

basically, it is working fine, but it fails randomly for unknown reasons

i ll add the new option as a workaround for cases like yours and close this

for the errors, enable "retry on fail" so that it will try again when it randomly fails

**Comment by kylon** [comment]
i think i found the way to reproduce the response error
[Sharedownt11.zip](https://github.com/kylon/Sharedown/files/10015445/Sharedownt11.zip)

this should fix it

**Comment by Elvino00** [comment]
> i think i found the way to reproduce the response error [Sharedownt11.zip](https://github.com/kylon/Sharedown/files/10015445/Sharedownt11.zip)
> 
> this should fix it

Do I have to use basic manual login?

**Comment by kylon** [comment]
yes

**Comment by Elvino00** [comment]
> yes

Sometimes I get the error "This page is not available, please check back in a few minutes" but I don't think it's sharedown's problem.

**Comment by Elvino00** [comment]
Every time I have to start the third download i get [the following error on the browser](https://user-images.githubusercontent.com/74961430/202169945-9bbdd09e-6f2a-4135-ad83-541aec25ad19.png). At this point I think there might be a problem.


[ytdlp.log](https://github.com/kylon/Sharedown/files/10021161/ytdlp.log)
[sharedownLog.log](https://github.com/kylon/Sharedown/files/10021162/sharedownLog.log)

**Comment by kylon** [comment]
yes, it will be slow, 2-3 should be fine, you can raise it when they stop limiting you

**Comment by Elvino00** [comment]
> yes, it will be slow, 2-3 should be fine, you can raise it when they stop limiting you

i guess than, apart from the throttling the problems are solved
